nano/node/websocket.cpp

```diff
@@ -0,0 +1,310 @@
+#include <algorithm>
+#include <boost/property_tree/json_parser.hpp>
+#include <chrono>
+#include <nano/node/node.hpp>
+#include <nano/node/websocket.hpp>
+
+nano::websocket::session::session (nano::websocket::listener & listener_a, boost::asio::ip::tcp::socket socket_a) :
+ws_listener (listener_a), ws (std::move (socket_a)), write_strand (ws.get_executor ())
+{
+	ws.text (true);
+}
+
+nano::websocket::session::~session ()
+{
+	ws_listener.get_node ().logger.try_log ("websocket session ended");
+}
+
+void nano::websocket::session::handshake ()
+{
+	std::lock_guard<std::mutex> lk (io_mutex);
+	ws.async_accept ([self_l = shared_from_this ()](boost::system::error_code const & ec) {
+		if (ec)
+		{
+			self_l->ws_listener.get_node ().logger.always_log ("websocket handshake failed: ", ec.message ());
+		}
+		else
+		{
+			// Start reading incoming messages
+			self_l->read ();
+		}
+	});
+}
+
+void nano::websocket::session::close ()
+{
+	std::lock_guard<std::mutex> lk (io_mutex);
+	boost::beast::websocket::close_reason reason;
+	reason.code = boost::beast::websocket::close_code::normal;
+	reason.reason = "Shutting down";
+	boost::system::error_code ec_ignore;
+	ws.close (reason, ec_ignore);
+}
+
+void nano::websocket::session::write (nano::websocket::message message_a)
+{
+	// clang-format off
+	std::lock_guard<std::mutex> lk (subscriptions_mutex);
+	if (message_a.topic == nano::websocket::topic::ack || subscriptions.find (message_a.topic) != subscriptions.end ())
+	{
+		subscriptions_mutex.unlock ();
```

Comment:
This should be a unique_lock::unlock instead of a lock_guard and then manually unlocking the mutex. When unlock is called and then lk goes out of scope, effectively unlock is called again which is UB.
https://en.cppreference.com/w/cpp/thread/mutex/unlock

Reply:
Ah, yeah that was bonkers. Thanks for spotting.